File: Design-Patterns/JS/Factory Pattern/Payment Factory/paymentFactory.js
class PaymentFactory {
  static registry = new Map();

  static register(type, clazz) {
    if (!type || typeof clazz !== "function") {
      throw new Error("Invalid registration");
    }

    if (this.registry.has(type)) {
      throw new Error(`Type ${type} already registered`);
    }

    this.registry.set(type, clazz);
  }

  static create(type, ...args) {
    const Clazz = this.registry.get(type);

    if (!Clazz) {
      throw new Error(`Unknown payment type: ${type}`);
    }

    return new Clazz(...args);
  }
}

module.exports = PaymentFactory;

File: Design-Patterns/Go/Factory-Pattern/Payment-Factory/paymentFactory.go
package main

import "sync"

var (
	registry = make(map[string]func(...any) Payment)
	mu       sync.RWMutex
)

// 3. Register function
func Register(name string, creator func(...any) Payment) {
	mu.Lock()
	defer mu.Unlock()

	if _, exists := registry[name]; exists {
		panic("type already registered: " + name)
	}

	registry[name] = creator
}

// 4. Factory
func Create(name string, args ...any) Payment {
	mu.RLock()
	creator, exists := registry[name]
	mu.RUnlock()

	if !exists {
		panic("unknown type: " + name)
	}

	return creator(args...)
}
